ghcide/src/Development/IDE/GHC/ExactPrint.hs

```diff
@@ -358,7 +471,7 @@ annotateDecl :: DynFlags -> LHsDecl GhcPs -> TransformT (Either String) (Anns, L
 annotateDecl dflags ast = do
     uniq <- show <$> uniqueSrcSpanT
     let rendered = render dflags ast
-    (anns, expr') <- lift $ mapLeft show $ parseDecl dflags uniq rendered
+    (anns, expr') <- lift $ mapLeft show $ parseDecls dflags uniq rendered
```

Comment:
Your `parseDecls` function seems to be an alternative version of [addAnnotationsForPretty](https://hackage.haskell.org/package/ghc-exactprint-0.6.3.4/docs/Language-Haskell-GHC-ExactPrint.html#v:addAnnotationsForPretty), which is specifically designed to add annotations to an AST element so that it is guaranteed to then parse again as the same AST.  It may change layout in the process, but there are extensive tests to ensure that it does work for all ASTs.

Reply:
Unfortunately `addAnnotationsForPretty` doesn't work for newly generated code: https://github.com/alanz/ghc-exactprint/issues/91